src/lib/server/lib.h

```diff
@@ -0,0 +1,73 @@
+#pragma once
+/*
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+/**
+ * $Id$
+ *
+ * @file lib/server/lib.h
+ * @brief API for initialising and freeing libraries.
+ *
+ * @copyright 2022 The FreeRADIUS server project
+ */
+RCSIDH(lib_h, "$Id$")
+
+#include <freeradius-devel/server/cf_parse.h>
+#include <freeradius-devel/util/dl.h>
+
+typedef int (*fr_lib_init_t)(void);
+
+typedef void (*fr_lib_free_t)(void);
+
+/** Structure to define how to initialise libraries with global configuration
+ *
+ */
+typedef struct {
+	char const 		*name;			//!<  Name of library and section within global config
+	CONF_PARSER const	*config;		//!<  Config parser for this library's global options
+	void			*inst;			//!<  Module data to parse global config into
+	fr_lib_init_t		init;			//!<  Callback to initialise library
+	fr_lib_free_t		free;			//!<  Callback to free library
+} fr_lib_autoinit_t;
+
+/*
+ *	To be used as terminator in an array of fr_lib_autoinit_t
+ */
+extern const fr_lib_autoinit_t fr_lib_autoinit_terminator;
+
+/** Structure to track use of libraries.
+ *
+ */
+typedef struct {
+	char const		*name;			//!<  Name of library
+	fr_rb_node_t		entry;			//!<  Entry in tree of libraries
+	void			*data;			//!<  Library configuration data
+	fr_lib_autoinit_t	*autoinit;		//!<  Autoinit structure used to manage this library
+	uint32_t		instance_count;		//!<  Number of current uses of this library
+	bool			initialised;		//!<  Has the init callback been run for this library
+} fr_lib_t;
+
+typedef struct {
+	fr_rb_tree_t		libs;
+} fr_lib_list_t;
+
+int fr_dl_lib_autoinit(dl_t const *module, void *symbol, void *user_ctx);
```

Comment:
Don't need the fr_ prefix that's only for functions in the util library and the protocol libraries, i.e. stuff that needs to be namespaced.  The server library is there so we can call the same code from the test harness binaries and `radiusd`, it's not meant for external consumption in the same way.